### PR TITLE
Improve set operations

### PR DESCRIPTION
**Pull request by negator**
Improves and adds operations on sets. Previously, there was no support for single ints or floats, since the aws sdk requires special handling for sets: http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.UpdateExpressions.html#Expressions.UpdateExpressions.ADD

In particular it requires the value that is being added or deleted, be explicitly set to a Set type.

### CODE REVIEW

File: domino.go [hunk 1]
```diff
@@ -52,16 +52,34 @@ type ToValue interface {
...
+	if len(m) <= 0 {
+		return
+	}
+	o = map[string]*dynamodb.AttributeValue{}
+	for k, v := range m {
+		switch t := v.(type) {
+		case *dynamodb.AttributeValue:
+			o[k] = t
+		default:
+			a, _ := dynamodbattribute.Marshal(t)
```
Comment (brendanjryan) [inline domino.go]: what will be set in the case of an error?

File: domino.go [hunk 2]
```diff
@@ -891,18 +921,13 @@ func (d *UpdateInput) ReturnNone() *UpdateInput {
...
-		if d.input.ExpressionAttributeValues == nil {
-			d.input.ExpressionAttributeValues = make(DynamoDBValue)
-		}
-		for k, v := range ea {
-			d.input.ExpressionAttributeValues[k] = v
-		}
+
+		d.input.ExpressionAttributeNames = n
+
+		d.input.ExpressionAttributeValues = marshal(m)
```
Comment (brendanjryan) [inline domino.go]: 👍 this is much cleaner

File: domino_test.go [hunk 3]
```diff
@@ -269,10 +269,11 @@ func TestUpdateItem(t *testing.T) {
 			table.loginCount.Increment(1),
 			table.lastLoginDate.SetField(time.Now().UnixNano(), false),
 			table.registrationDate.SetField(time.Now().UnixNano(), true),
-			table.visits.Append(time.Now().UnixNano()),
-			table.preferences.RemoveKey("update_email"),
+			// table.visits.AddInteger(time.Now().UnixNano()),
```
Comment (brendanjryan) [inline domino_test.go]: nit: remove

File: domino_test.go [hunk 4]
```diff
@@ -269,10 +269,11 @@ func TestUpdateItem(t *testing.T) {
...
+			table.preferences.Remove("update_email"),
 		)
 
+	fmt.Println(u.Build())
```
Comment (brendanjryan) [inline domino_test.go]: same

File: expression.go [hunk 5]
```diff
@@ -92,26 +109,26 @@ func And(c ...Expression) ExpressionGroup {
...
 }
 
 /*********************************************************************************/
 /******************************** Negation Expression ****************************/
 /*********************************************************************************/
 
-func (n negation) construct(counter uint, topLevel bool) (string, map[string]interface{}, uint) {
-	s, m, c := n.expression.construct(counter, topLevel)
+func (n negation) construct(counter uint, topLevel bool) (string, map[string]*string, map[string]interface{}, uint) {
+	s, names, m, c := n.expression.construct(counter, topLevel)
```
Comment (brendanjryan) [inline expression.go]: nit: in other similar places you kept this variable as `n`. not sure if it adds additional clarity here

File: expression.go [hunk 6]
```diff
@@ -268,67 +288,136 @@ func (Field *DynamoField) SetField(a interface{}, onlyIfEmpty bool) *UpdateExpre
...
-		m := make(map[string]interface{})
-		return s, m, c
+		return s, nil, m, c
+	}
+	return &UpdateExpression{op: "SET", f: f}
+}
+
+func (Field *dynamoListField) Remove(index int) *UpdateExpression {
+	f := func(c uint) (string, map[string]*string, map[string]interface{}, uint) {
+		s := fmt.Sprintf("%v[%v]", Field.name, index)
```
Comment (brendanjryan) [inline expression.go]: nit: we could probably use more specific format strings when possible

File: expression.go (hunk 5 above)
Comment (negator) [inline expression.go]: `n` clashed with the receiver name

File: domino.go (hunk 1 above)
Comment (negator) [inline domino.go]: Unfortunately, this gets invoked during the construction of a query in the dsl. Propagating the error back would require complex delayed function logic that we have in other spots. For now I've chosen to ignore them, but probably better to panic. This should never throw however.

**Pull request merged**